ccmbioinfo/lfs
# CARDS-1424: Send email to patients 72 hours prior to appointment, prompting them to complete the surveys

This PR implements CARDS-1424 by providing a scheduled task that checks if any patients have any appointments on the third day from the day that the task executes. For example, if the task executes on _2022-01-01_ at _01:30:00_ it will check for any appointments happening at any time on _2022-01-04_. For each appointment that is found, the task will check if the patient has provided their email address and has consented to email communications. If they have, they will receive an email notifying them of their upcoming appointment and prompting them to complete their surveys.

To test:

- Follow the instructions at https://github.com/data-team-uhn/cards/pull/822#issue-1063852994 to configure a local debugging mail server for CARDS to use.
- Start CARDS with the command `CARDS_HOST_AND_PORT='datapro.uhn.ca' CLINIC_SLING_PATH='/Proms.html/Cardio' PATIENT_NOTIFICATION_FROM_ADDRESS='datapro@uhn.ca' PATIENT_NOTIFICATION_FROM_NAME='UHN DATAPRO' NIGHTLY_NOTIFICATIONS_SCHEDULE='0 * * * * ? *' SLING_COMMONS_CRYPTO_PASSWORD=password ./start_cards.sh --project cards4proms --dev -f mvn:io.uhndata.cards/cards-email-notifications/0.9-SNAPSHOT/slingosgifeature -V emailnotifications.smtps.port=8465 -V emailnotifications.smtps.host=localhost`.
- Login to CARDS as `admin`:`admin`.
- Create a new _Patient information_ Form and, with it, a new _Patient_ subject (`P1`).
- Complete the _Patient information_ and provide an email address and select `Yes` for the question `The patient consents to be contacted by email`.
- Create a new _Visit information_ Form and, with it, a new _Visit_ subject (`V1`). Select `P1` as the _parent Patient_ for this `V1` subject.
- Set the _Time_ for this Form to be in three days from now. This step will not work in Firefox but will work in Chrome.
- Save the Form.
- Open Composum at `http://localhost:8080/bin/browser.html`
- Make note of the `jcr:uuid` value for the `/Questionnaires/Visit information/surveys` node.
- Find the JCR node under `/Forms/` that has a `question` value equal to the `jcr:uuid` value from the previous step.
- Add a _string_ property named `value` with the value of `Cardio` - here, we are setting the value _Cardio_ as the response to the _Surveys_ question for this _Visit information_ form.
- In the console with Python's SMTP debugging server, you should see email notifications being received. The rate at which the notifications are received is set by the `NIGHTLY_NOTIFICATIONS_SCHEDULE` environment variable which, for testing purposes, I have provided instructions so that it runs at the start of every minute.
- Ensure that the link works and visiting it (replacing `datapro.uhn.ca` with `localhost:8080` and `https` with `http`) in a browser window that is not logged into CARDS works as expected.
- Experiment with different patients with different scheduled visits at different clinics and ensure that everything works as expected.
